File: snake-scala-2d/src/main/scala/metacake/scala/snake/entity/Snake.scala
package metacake.scala.snake.entity

import java.awt.Color
import io.metacake.s2d.output.drawing.instructions._
import metacake.scala.snake.entity.Direction.Direction

object Snake {
  val COLOR: Color = Color.GREEN
  val SIZE: Int = 50
  val IMAGE: DrawInstruction = new RectangleInstruction(Snake.SIZE, Snake.SIZE, Snake.COLOR)
}

object Direction extends Enumeration {
  type Direction = Value
  val UP, DOWN, LEFT, RIGHT = Value
}

class Segment(val x: Int, val y: Int) extends Positionable {
  override def equals(o: Any): Boolean = o match {
    case that: Segment => this.x == that.x && this.y == that.y
    case _ => false
  }
  override def toString: String = "(" + x +", " + y + ")"
}

class Snake(val dir: Direction, val segments: List[Segment]) extends Drawable {

  def move(): Snake = new Snake(dir, nextHead() :: segments.dropRight(1))

  def grow(): Snake = {
    def tail: Segment = segments.last
    new Snake(dir, (nextHead() :: segments.dropRight(1)) :+ new Segment(tail.x, tail.y))
  }

  private def nextHead(): Segment = dir match {
    case Direction.UP => headInDirection(0, -1)
    case Direction.DOWN => headInDirection(0, 1)
    case Direction.LEFT => headInDirection(-1, 0)
    case Direction.RIGHT => headInDirection(1, 0)
  }

  private def headInDirection(x: Int, y: Int): Segment = {
    def head = segments.head
    new Segment(x + head.x, y + head.y)
  }

  override def draw(scene: DrawInstruction): DrawInstruction = {
    def offset(pos: Int): Int = pos * Snake.SIZE
    segments.foldLeft(scene)({(acc: DrawInstruction, segment: Segment) =>
      new PlaceInstruction(Snake.IMAGE, offset(segment.x), offset(segment.y), acc)
    })
  }

  override def equals(o: Any): Boolean = o match {
    case that: Snake => this.segments == that.segments
    case _ => false
  }
  override def toString: String = segments.toString()
}

File: java-snake-2d/src/main/java/metacake/snake/entity/Snake.java
package metacake.snake.entity;

import io.metacake.s2d.output.drawing.instructions.DrawInstruction;

import java.util.LinkedList;

public class Snake implements Drawable {

    public static Snake initialSnake(Direction direction) {
        LinkedList<SnakeSegment> head = new LinkedList<>();
        head.add(new SnakeSegment(0, 0));
        return new Snake(direction, head);
    }

    private Direction direction;
    private LinkedList<SnakeSegment> segments;

    Snake(Direction direction, LinkedList<SnakeSegment> segments) {
        this.direction = direction;
        this.segments = segments;
    }

    @Override
    public DrawInstruction draw(DrawInstruction instruction) {
        DrawInstruction image = instruction;
        for (SnakeSegment segment : segments) {
            image = segment.draw(image);
        }
        return image;
    }

    public void setDirection(Direction direction) {
        this.direction = direction;
    }

    public void move() {
        SnakeSegment snakeHead = segments.getFirst();
        SnakeSegment head = new SnakeSegment(direction.getChangeX() + snakeHead.getX(), direction.getChangeY() + snakeHead.getY());
        segments.addFirst(head);
        segments.removeLast();
    }

    public void eat() {
        SnakeSegment snakeHead = segments.getFirst();
        SnakeSegment head = new SnakeSegment(direction.getChangeX() + snakeHead.getX(), direction.getChangeY() + snakeHead.getY());
        segments.addFirst(head);
    }

    public boolean canEat(Food food) {
        SnakeSegment snakeHead = segments.getFirst();
        return snakeHead.getX() == food.getX() && snakeHead.getY() == food.getY();
    }

    public boolean isEatingSelf() {
        SnakeSegment snakeHead = segments.getFirst();
        return segments.stream().skip(1).anyMatch((segment) -> segment.equals(snakeHead));
    }

    @Override
    public String toString() {
        return segments.toString();
    }

    public int getHeadX() {
        return segments.getFirst().getX();
    }

    public int getHeadY() {
        return segments.getFirst().getY();
    }

    public int size() {
        return segments.size();
    }
}